### PR TITLE
Gcp/configurable ip alias ranges

### PR DESCRIPTION
**Pull request by brandonbell**
### Summary
Make AliasIPRange configurable at the cluster and instanceGroup scope defaulting to /24.

### CODE REVIEW

File: pkg/model/gcemodel/context.go [hunk 1]
```diff
@@ -132,6 +132,14 @@ func (c *GCEModelContext) NameForFirewallRule(id string) string {
 	return gce.ClusterSuffixedName(id, c.Cluster.ObjectMeta.Name, 63)
 }
 
+func (c *GCEModelContext) NetworkIPAliasRange() *string {
+	if c.Cluster.Spec.CloudConfig.GCPAliasRange != nil {
+		return c.Cluster.Spec.CloudConfig.GCPAliasRange
+	}
```
Comment (gwynethallwright) [inline pkg/model/gcemodel/context.go:138]: Are we validating that the string value is sensible anywhere?

Comment (brandonbell) [reply pkg/model/gcemodel/context.go:138]: Thanks, I can add that!

File: k8s/crds/kops.k8s.io_clusters.yaml [hunk 2]
```diff
@@ -390,6 +390,10 @@ spec:
                     description: GCEServiceAccount specifies the service account with
                       which the GCE VM runs
                     type: string
+                  gcpAliasRange:
+                    description: GCPAliasRange: Alias range from the pods CIDR that will be 
+                      assigned to the instances, defaults to /24
+                    type: string
```
Comment (cmparkinson) [inline k8s/crds/kops.k8s.io_clusters.yaml:396]: I'm not familiar with how strict the schema is and if you can convert types later, but I have question / idea.

Does using `integer` here simplify things at all?  If you leave this as a `string`, validation will require that you strip the `/`, convert to `int`, then do a bounds check on the number of bits in the mask to enforce that it's within a reasonable range.  If you start with an `integer` field, you can skip the first two steps and jump right to using comparison operators, then simply `fmt.Sprintf("/%d", variableNameHere)` to get the string representation later.

Comment (cmparkinson) [reply k8s/crds/kops.k8s.io_clusters.yaml:396]: Building on this, we _might_ be able to use OpenAPI v3 schema validations and defaulting to do everything for us (the unquoted colon in the description looks a bit off too)  Something like:
```yaml
              gcpAliasRange:
                description: Alias range from the pods CIDR that will be
                  assigned to the instances, defaults to 24
                type: integer
                minimum: 24
                maximum: 28
                default: 24
```

Comment (cmparkinson) [reply k8s/crds/kops.k8s.io_clusters.yaml:396]: Here's an example that uses `kubebuilder` annotations to generate the validation rules: https://github.com/Invoca/kops/commit/38bb3388e8a9e188c034393b85aa1f73d093b9dc

Comment (brandonbell) [reply k8s/crds/kops.k8s.io_clusters.yaml:396]: Doesn't seem these annotations actually work.   I've added validation to https://github.com/Invoca/kops/blob/c6fce4eafcbe40f416de95a183910fc8354c85f7/pkg/apis/kops/validation/gce.go

File: pkg/model/gcemodel/context.go (hunk 1 above)
Comment (brandonbell) [reply pkg/model/gcemodel/context.go:138]: Added in https://github.com/Invoca/kops/blob/c6fce4eafcbe40f416de95a183910fc8354c85f7/pkg/apis/kops/validation/gce.go

File: pkg/apis/kops/validation/gce.go [hunk 3]
```diff
@@ -44,15 +46,32 @@ func gceValidateCluster(c *kops.Cluster) field.ErrorList {
...
-		fieldSpec := field.NewPath("spec")
 		allErrs = append(allErrs, IsValidValue(fieldSpec.Child("gcpProvisioningModel"), ig.Spec.GCPProvisioningModel, []string{"STANDARD", "SPOT"})...)
 	}
+
+	gcpAliasRange := ig.Spec.GCPAliasRange
+	if !gceValidateAliasRange(gcpAliasRange) {
+		allErrs = append(allErrs, field.Invalid(fieldSpec.Child("gcpAliasRange"), gcpAliasRange, "must match /2[4-8]"))
+	}
 	return allErrs
 }
```
Comment (brandonbell) [inline pkg/apis/kops/validation/gce.go]: Note, validation doesn't seem to run for `instanceGroups` when doing a `kops replace -f` but it does for `kops create -f ` and `kops update instancegroup <ig>` .  This seems like a bug outside the scope of this change.  

`kops replace -f` on the cluster.yaml does run through validation.

File: pkg/model/gcemodel/context.go (hunk 1 above)
Comment (gwynethallwright) [reply pkg/model/gcemodel/context.go:138]: Nice!

**Comment by gwynethallwright** [comment]
Are the CI failures expected?

**Comment by brandonbell** [comment]
@gwynethallwright Thanks, they're not!  

Looks like the validation step needs to be gated for `network: gce` only.  Thanks!

**Review by cmparkinson — APPROVED** [review]
LGTM 👍

**Comment by brandonbell** [comment]
@gwynethallwright I have the Ci tests passing now and I'm going to merge.  Thanks!

**Pull request merged**